src/System.Buffers/src/System/Buffers/DefaultArrayPool.cs

```diff
@@ -0,0 +1,151 @@
+            _numberOfArraysPerBucket = arraysPerBucket;
+            _buckets = new ArrayPoolBucket<T>[maxBuckets + 1];
+            _rentedBufferCount = 0;
+            _totalBufferCount = 0;
+        }
+
+        public override T[] Rent(int length)
+        {
+            if (length <= 0)
+                throw new ArgumentOutOfRangeException(nameof(length));
+
+            Interlocked.Increment(ref _rentedBufferCount);
+
+            int index = Utilities.SelectBucketIndex(length);
+            if (index >= _buckets.Length)
+            {
+                Interlocked.Increment(ref _totalBufferCount);
+                return new T[length];
+            }
+            else
+            {
+                T[] buffer = null;
+
+                // Search for an array starting at the 'index' bucket. If the bucket
+                // is empty, bump up to the next higher bucket and try that one
+                for (int i = index; i < _buckets.Length; i++)
+                {
+                    // Check if the bucket has been created and, if not, create it
+                    ArrayPoolBucket<T> bucket = Volatile.Read(ref _buckets[i]);
+                    if (bucket == null)
+                    {
+                        if (Interlocked.CompareExchange(ref _buckets[index], new ArrayPoolBucket<T>(Utilities.GetMaxSizeForBucket(index), _numberOfArraysPerBucket), null) == null)
+                        {
+                            Interlocked.Add(ref _totalBufferCount, _numberOfArraysPerBucket);
+                        }
+                        bucket = _buckets[index];
+                    }
+
+                    buffer = bucket.Rent();
+
+                    // If the bucket has an array left and returned it, give it to the caller
+                    if (buffer != null)
+                    {
+                        return buffer;
+                    }
+                }
+
+                // Getting here means that all Buckets from index through _buckets.Length are taken
+                // so we need to allocate a buffer on demand
+                Interlocked.Increment(ref _totalBufferCount);
+                return new T[length];
+            }
+        }
+
+        public override T[] Enlarge(T[] original, int newLength)
+        {
+            if (original == null)
+                throw new ArgumentNullException(nameof(original));
+            // Our buffers have a minimum size of 16, so check for that first
+            if ((newLength > MinimiumArraySize || original.Length > MinimiumArraySize) &&
+                (newLength < original.Length))
+                throw new ArgumentOutOfRangeException(nameof(newLength), SR.Invalid_Enlarge_New_Size);
+
+            // Since we can't realloc, get a bigger buffer, copy the data, clear the old buffer, and return the new one
+            T[] newBuffer = Rent(newLength);
+            Array.Copy(original, newBuffer, original.Length);
+            Return(original, clearArray: true);
```

Comment:
I still question having this Enlarge method exposed at all.  A developer can write these three lines on their own.  Exposing it seems a bit misleading, since we can't actually change the size of the array.  And if a developer does need to enlarge like this, they can write these three lines themselves, and without all of the above error checking we need to do.  And they'd get to decide on the value they want for clearArray.  We could always add this method or something like it back in the future if it's really needed.